sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,27 +10,229 @@
 
 typedef struct {
   struct curl_slist * p_list;
-} az_headers_data;
+} az_curl_headers_list;
 
-AZ_CALLBACK_FUNC(az_headers_to_curl, az_headers_data *, az_pair_visitor)
+/**
+ * @brief writes a header key and value to a buffer as a 0-terminated string and using a separator
+ * span in between. Returns error as soon as any of the write operations fails
+ *
+ * @param writable_buffer
+ * @param p_header
+ * @param separator
+ * @return az_result
+ */
+az_result az_write_to_buffer(
+    az_span const writable_buffer,
+    az_pair * const p_header,
+    az_const_span const separator) {
+  az_span_builder writer = az_span_builder_create(writable_buffer);
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, p_header->key));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, separator));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, p_header->value));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, AZ_STR("\0")));
+  return AZ_OK;
+}
+
+/**
+ * @brief allocate a buffer for a header. Then reads the az_pair header and writes a buffer. Then
+ * uses that buffer to set curl header. Header is set only if write operations were OK. Buffer is
+ * free after setting curl header.
+ *
+ * @param p_header
+ * @param p_headers
+ * @param separator
+ * @return az_result
+ */
+az_result az_add_header_to_curl_list(
+    az_pair * const p_header,
+    az_curl_headers_list * const p_headers,
+    az_const_span separator) {
+  // allocate a buffet for header
+  int16_t const buffer_size = p_header->key.size + separator.size + p_header->value.size + 1;
+  uint8_t * const p_writable_buffer = (uint8_t * const)malloc(buffer_size);
+  if (p_writable_buffer == NULL) {
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+  char * buffer = (char *)p_writable_buffer;
+
+  // write buffer
+  az_span const writable_buffer
+      = (az_span const){ .begin = p_writable_buffer, .size = buffer_size };
+  az_result const write_result = az_write_to_buffer(writable_buffer, p_header, separator);
+
+  // attach header only when write was OK
+  if (az_succeeded(write_result)) {
+    p_headers->p_list = curl_slist_append(p_headers->p_list, buffer);
+  }
+  // at any case, error or OK, free the allocated memory
+  free(p_writable_buffer);
+  return write_result;
+}
+
+/**
+ * @brief loop all the headers from a HTTP request and set each header into easy curl
+ *
+ * @param p_hrb
+ * @param p_headers
+ * @return az_result
+ */
+az_result az_build_headers(
+    az_http_request_builder * const p_hrb,
+    az_curl_headers_list * p_headers) {
+  az_const_span separator = AZ_STR(": ");
```

Comment:
Don't keep it on stack. Use either, `static const` or inlined. We don't mutate the separator.